### PR TITLE
Add new page for Google BigQueryIO

### PR DESCRIPTION
**Pull request by melap**
Do not merge yet: This PR is still in progress alongside the snippets in accompanying PR apache/beam#4363

### CODE REVIEW

**Comment by melap** [comment]
also CC: @davidcavazos

This is missing a couple snippets still (and I'll need to change the snippet paths once that PR is merged), but starting a review on the bulk of it.

**Comment by chamikaramj** [comment]
cc: @reuvenlax

File: src/documentation/io/built-in-google-bigquery.md [hunk 1]
```diff
@@ -0,0 +1,753 @@
...
+
+
+<!-- Java specific -->
+
+{:.language-java}
+The Beam SDK for Java has two BigQueryIO read methods. Both of these methods
+allow you to read from a table, or read fields using a query string.
+
+{:.language-java}
+1. `readTableRows` returns a `PCollection` of BigQuery `TableRow`
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: I'd recommend to swap the order of these two methods because the second is typically preferable.

File: src/documentation/io/built-in-google-bigquery.md [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+When writing to BigQuery, you must supply a table schema for the fields you want
+to write to the target table unless you specify a [create disposition]() of
+`CREATE_NEVER`. [Creating a table schema](#creating-a-table-schema) covers
+schemas in more detail.
+
+
+## Reading from BigQuery
+
+BigQueryIO allows you to read from a BigQuery table, or read specific fields
+using a query string. When you apply a BigQueryIO read transform in batch mode,
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: Remove "in batch mode" - Beam doesn't have modes (only bounded or unbounded PCollection's) and the read transform does not have separate bounded/unbounded codepaths (it's not applied to a PCollection, but to the root of the pipeline).

Dataflow runner does have modes, and using BigQueryIO.read() is equally valid in batch and streaming Dataflow runner, and it works the same way.

File: src/documentation/io/built-in-google-bigquery.md [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+
+{:.language-py}
+To read an entire BigQuery table, use the `table` parameter with the BigQuery
+table name.
+
+The following code reads an entire table that contains weather station data and
+then extracts the `max_temperature` column.
+
+```java
+{% github_sample /davidcavazos/beam/blob/bigquery-snippets/examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java tag:BigQueryReadTable
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: This should probably use the snippet that uses read(SerializableFunction) - I think David has created those snippets already. Same for the other snippets.

File: src/documentation/io/built-in-google-bigquery.md [hunk 4]
```diff
@@ -0,0 +1,753 @@
...
+`read(SerializableFunction)`.
+
+{:.language-py}
+If you don't want to read an entire table, you can supply a query string to
+`BigQuerySource` by specifying the `query` parameter.
+
+The following code uses a SQL query to only read the `max_temperature` column.
+
+```java
+{% github_sample /davidcavazos/beam/blob/bigquery-snippets/examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java tag:BigQueryReadQuery
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: @davidcavazos This snippet uses `SerializableCoder.of(Double.class)` - that's a very inefficient coder and should use `DoubleCoder.of()` instead.

File: src/documentation/io/built-in-google-bigquery.md [hunk 5]
```diff
@@ -0,0 +1,753 @@
...
+   should create a new table if one does not exist. If you use this value, you
+   must provide a table schema. `CREATE_IF_NEEDED` is the default behavior.
+
+ * `BigQueryDisposition.CREATE_NEVER`: Specifies that a table should never be
+   created. If the target table does not exist, the write operation fails.
+
+
+If you specify `CREATE_IF_NEEDED` as the create disposition and you don't
+supply a table schema, the transform may fail at runtime with a
+<span class="language-java">java.lang.IllegalArgumentException</span>
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: It's not guaranteed that it will be this particular exception. Just say that it may fail.

File: src/documentation/io/built-in-google-bigquery.md [hunk 6]
```diff
@@ -0,0 +1,753 @@
...
+streaming inserts. Each insertion method provides different tradeoffs of cost,
+quota, and data consistency. See the BigQuery documentation for
+[load jobs](https://cloud.google.com/bigquery/loading-data) and
+[streaming inserts](https://cloud.google.com/bigquery/streaming-data-into-bigquery)
+for more information about these tradeoffs.
+
+BigQueryIO chooses a default method based on the input `PCollection`.
+
+* Beam uses a BigQuery load job when you apply a BigQueryIO write transform in
+  batch mode to write to a single table.
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: This doesn't sound right. It uses load jobs (plural - it may use multiple load jobs per table if the amount of data exceeds BigQuery limits for a single load job) if you apply it to a bounded `PCollection`, or if you use `BigQueryIO.write().withTriggeringFrequency()` (which is worth listing as a separate method).

File: src/documentation/io/built-in-google-bigquery.md [hunk 7]
```diff
@@ -0,0 +1,753 @@
...
+
+* Beam uses BigQuery streaming inserts when you apply a BigQueryIO write
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: This is no longer true - see previous comments. Streaming inserts are used by default only for unbounded collections unless triggering frequency or an explicit method is specified.

File: src/documentation/io/built-in-google-bigquery.md [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+```
+
+
+### Using dynamic destinations
+
+{:.language-py}
+> The Beam SDK for Python does not currently support dynamic destinations.
+
+The dynamic destinations feature allows you to specify dynamic BigQuery table
+destinations when you write your `TableRow` objects. Dynamic destinations groups
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: I'd rephrase the first sentence: ...allows you to write different elements of a PCollection to different BigQuery tables, possibly with different schemas.

File: src/documentation/io/built-in-google-bigquery.md [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+your user type by a key (or destination), and writes each group to a different
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: "or" -> "i.e. destination", or "called "destination"" or something like that, because these are not alternative options but synonyms/clarifications

File: src/documentation/io/built-in-google-bigquery.md [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+
+## Limitations
+
+BigQueryIO currently has the following limitations.
+
+1. You can’t sequence the completion of a BigQuery write with other steps of
+   your pipeline.
+
+2. There is no efficient way to import a large time-partitioned table. As a
+   workaround, you can use dynamic destinations with a table per partition.
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: With a table decorator per partition

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

File: src/documentation/io/built-in-google-bigquery.md (hunk 9 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: I tinkered with this a bit, PTAL -- does "destination key" work?

File: src/documentation/io/built-in-google-bigquery.md (hunk 5 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Done

File: src/documentation/io/built-in-google-bigquery.md (hunk 2 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Done

File: src/documentation/io/built-in-google-bigquery.md (hunk 1 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Done

File: src/documentation/io/built-in-google-bigquery.md (hunk 8 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Done

File: src/documentation/io/built-in-google-bigquery.md (hunk 10 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Done

File: src/documentation/io/built-in-google-bigquery.md [hunk 11]
```diff
@@ -561,9 +568,10 @@ The following example code shows how to apply a `Write` transform to write a
 table if needed; if the table already exists, it will be replaced.
 
 ```java
-{% github_sample /davidcavazos/beam/blob/bigquery-snippets/examples/java8/src/main/java/org/apache/beam/examples/snippets/Snippets.java tag:BigQueryWriteFunction
+{% github_sample /apache/beam/blob/master/examples/java/src/main/java/org/apache/beam/examples/snippets/Snippets.java tag:BigQueryWriteFunction
 %}```
 ```py
+# This Python sample coming soon.
```
Comment (melap) [inline src/documentation/io/built-in-google-bigquery.md]: Found one missing snippet in snippet file  @davidcavazos

**Comment by melap** [comment]
@chamikaramj can you also take a quick look at this for the Python version?

File: src/documentation/io/built-in-google-bigquery.md (hunk 11 above)
Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: from my understand, we recommend Write for batch until https://issues.apache.org/jira/browse/BEAM-3289 is finished and it will just do the best thing, so I think we should have some snippet that uses Write even if it's different

Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: apache/beam#4771

File: src/documentation/io/built-in-google-bigquery.md [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+### Writing to a table
+
+{:.language-java}
+To write to a BigQuery table, apply either a `writeTableRows` or `write`
+transform.
+
+{:.language-py}
+To write to a BigQuery table, apply either a `Write` or `WriteToBigQuery` transform.
+`Write` supports batch mode. `WriteToBigQuery` supports both batch mode and
+streaming mode, but it is less efficient for batch mode. You must apply the
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: "less efficient for batch mode" this is not correct since Python DataflowRunner replaces WriteToBigQuery with Write at runtime. I think you can only mention WriteToBigQuery here (which we will be update with a custom BQ sink hopefully in the near future). We will be adding a ReadFromBigQuery (custom BQ source) as well but we can update the documentation once we have that and when we want to make it the default.

File: src/documentation/io/built-in-google-bigquery.md [hunk 13]
```diff
@@ -0,0 +1,761 @@
...
+
+
+## Limitations
+
+BigQueryIO currently has the following limitations.
+
+1. You can’t sequence the completion of a BigQuery write with other steps of
+   your pipeline.
+
+2. There is no efficient way to import a large time-partitioned table. As a
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: I don't think this is true after https://github.com/apache/beam/pull/4694.

File: src/documentation/io/built-in-google-bigquery.md [hunk 14]
```diff
@@ -0,0 +1,761 @@
...
+   workaround, you can use dynamic destinations with a table decorator per
+   partition. However, this can create a lot of load jobs and you may run out
+   of quota (1000 jobs/day).
+
+3. If you are using the Beam SDK for Python, you may have import size quota
+   issues if you write a very large data set. As a workaround, you can partition
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: partition the dataset (for example using Beam's Partition) transform and write to multiple BigQuery tables.

File: src/documentation/io/built-in-google-bigquery.md (hunk 11 above)
Comment (chamikaramj) [reply src/documentation/io/built-in-google-bigquery.md]: There's no way to provide convert a custom type conversion function to Python (and I'm not sure if we need once since Python writes dictionaries). Following text and the snippet should be dropped from the Python doc.

"The following example code shows how to apply a Write transform to write a PCollection of dictionaries to a BigQuery table. The write operation creates a table if needed; if the table already exists, it will be replaced."

File: src/documentation/io/built-in-google-bigquery.md (hunk 12 above)
Comment (chamikaramj) [reply src/documentation/io/built-in-google-bigquery.md]: Here's the place where we perform transform replacement for DataflowRunner: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L483

Comment (melap) [reply src/documentation/io/built-in-google-bigquery.md]: Oh, okay, so bullet 2 in https://issues.apache.org/jira/browse/BEAM-3289 is done? in that case I will remove it and have WriteToBigQuery as the only option.

Comment (chamikaramj) [reply src/documentation/io/built-in-google-bigquery.md]: Yeah, it was already done. Added  a comment to the JIRA. Sorry about the confusion.

File: src/documentation/io/built-in-google-bigquery.md [hunk 15]
```diff
@@ -0,0 +1,761 @@
...
+
+{:.language-java}
+* `withTimePartitioning`: This method takes a `TimePartitioning` class, and is
+  only usable if you are writing to a single table.
+
+* `withJsonTimePartitioning`: This method is the same as
+  `withTimePartitioning`, but takes a JSON-serialized String object.
+
+{:.language-java}
+This example generates one partition per day.
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: Once 2.4 gets released we'll need to update this per https://github.com/apache/beam/pull/4694 because it is WAY more efficient and concise, and has to be very clearly called out as the right way to do time partitioning (which is likely the most common use case of dynamic destinations).

File: src/documentation/io/built-in-google-bigquery.md [hunk 16]
```diff
@@ -0,0 +1,761 @@
...
+   the data set. The Beam SDK for Java does not have this limitation as it
+   partitions your data set for you.
+
+
+## Additional examples
+
+You can find additional examples that use BigQuery in Beam's examples
+directories.
+
+### Java cookbook examples
```
Comment (jkff) [inline src/documentation/io/built-in-google-bigquery.md]: We should audit these examples for whether they comply to the currently known best practices...

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by melap** [comment]
retest this please

**Comment by swegner** [comment]
FYI, Beam Website source code is being migrated to [apache/beam](https://github.com/apache/beam), and we will retire the beam-site repository after all in-progress pull requests have been completed.

Is this PR still relevant? If so, do you have an ETA of when it can be merged?

**Comment by chamikaramj** [comment]
I'll review this week. Sorry about the delay.

**Comment by chamikaramj** [comment]
Retest this please

File: src/documentation/io/built-in-google-bigquery.md [hunk 17]
```diff
@@ -0,0 +1,749 @@
...
+
+{:.language-java}
+To use dynamic destinations, you must create a `DynamicDestinations` object and
+implement the following methods:
+
+{:.language-java}
+* `getDestination`: Returns an object that `getTable` and `getSchema` can use as
+  the destination key to compute the destination table and/or schema.
+
+* `getTable`: Returns the table (as a `TableDestination` object) for the
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: Please add: " This must return a unique table for each unique destination."

File: src/documentation/io/built-in-google-bigquery.md [hunk 18]
```diff
@@ -0,0 +1,749 @@
...
+[load jobs](https://cloud.google.com/bigquery/loading-data) and
+[streaming inserts](https://cloud.google.com/bigquery/streaming-data-into-bigquery)
+for more information about these tradeoffs.
+
+BigQueryIO chooses a default insertion method based on the input `PCollection`.
+
+BigQueryIO uses load jobs in the following situations:
+
+* When you apply a BigQueryIO write transform to a bounded `PCollection`.
+* When you apply a BigQueryIO write transform to an unbounded `PCollection` and
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: Last two points are Java only.

File: src/documentation/io/built-in-google-bigquery.md [hunk 19]
```diff
@@ -0,0 +1,749 @@
...
+  `BigQueryIO.write().withMethod(FILE_LOADS)`.
+
+BigQueryIO uses streaming inserts in the following situations:
+
+* When you apply a BigQueryIO write transform to an unbounded `PCollection` and
+  do not set the triggering frequency.
+* When you specify streaming inserts as the insertion method.
+
+When you use streaming inserts, you can decide what to do with failed records.
+You can either keep retrying, or return the failed records in a separate
```
Comment (chamikaramj) [inline src/documentation/io/built-in-google-bigquery.md]: This is Java only.

**Comment by melap** [comment]
Thanks! Made changes based on your feedback.

**Comment by melap** [comment]
@asfgit merge

**Comment by melap** [comment]
@asfgit

**Comment by melap** [comment]
@asfgit merge

**Comment by asfgit** [comment]
Error: PR failed in verification; check the Jenkins job for more information.

**Comment by swegner** [comment]
Verification job [failed](https://builds.apache.org/job/beam_PreCommit_Website_Merge/369) while checking `archive.apache.org/dist` links for release sources. Pinging one of them this morning works; I'll try one more time and if it still fails merge manually.

**Comment by swegner** [comment]
@asfgit merge
